### DISCUSSION TITLE
Trying to reuse base64 content instead of temporary file path

### DISCUSSION THREAD

**Discussion by VincentCATILLON** [react-native-image-picker/react-native-image-picker#1525] [Q&A]
Hi @ravirajn22 & everyone,

We are making an offline application that uses this lib from the 2.x version (upgraded to 3.0.1 few days ago) and encountered an issue when we pick a photo (camera or gallery) with your lib during offline period and retrying to post the photo later (ie: 10 seconds after, using `apollo-upload-client` in our case) pointing to the temporary file which is deleted automatically (maybe by the lib or with magic, but i know, mentioned in your [documentation](https://github.com/react-native-image-picker/react-native-image-picker#note-on-file-storage))

> Could not retrieve file for contentUri content://foo.bar.baz.imagepickerprovider/external/Android/data/foo.bar.baz/files/rn_image_picker_lib_temp_2b661297-00f8-44a1-9e4e-b2b1ffab6e98.jpg

Is there any way to get around this issue using our `AsyncStorage` to get the base64 content of the photo, decode it later just before trying to upload to ensure that we have the content and rely on controlled content ?

I tried to do this (for example with camera usage):

```js
import { launchCamera } from 'react-native-image-picker';

const handlePickFinish = useCallback(
  (response: ImagePickerResponse) => { 
    // [...] Check on response.errorCode, response.didCancel and response.uri before
    AsyncStorage.setItem('foo', response.base64);
  },
  []
);

const handlePress = useCallback(() => {
   // [...] Permission check/request before
   return launchCamera(CAMERA_OPTIONS, handlePickFinish);
});
```

and reuse it during upload like:

```js
import base64 from 'base-64';

export const getFile = async () => {
  const encodedContent = await AsyncStorage.getItem('foo');
  // Check on encodedContent !== null
  const content = base64.decode(encodedContent);
  
  return new Blob([content], { type: 'image/jpeg' });
};
```

But it doesn't work as expected.

Did i miss something to reuse base64 content of pictures ?

**Comment by ravirajn22** [comment]
@VincentCATILLON In iOS we use cache folder, so the OS will clear it but It will happen very rarely when space is low. On Android since there is no cache folder managed by android, so we delete the previous image/video file when the user uses the the library next time. So, until he does not use the library methods, the file is safe. So please check what you'r doing, better you check apollo-upload-client library if its handling URI properly.

Regarding using base64, use it only for small sized image, not large image, else your whole app will hang for few sec. String processing and RN bridge problem.  But uploading base64 should not have any trouble, I think you are having doing something wrong by decoding it. I don't know if the decoding is working properly for image file type. You check online. Find a way to directly send the base64 instead of  decoding in client.

**Comment by VincentCATILLON** [comment]
@ravirajn22 Firstly, thanks a lot for your (quick) answer and explanation, I really appreciate! 👍 

As mentioned in the post description, as our application is offline first, the user can pick multiple photos from gallery and/or using camera, and we retry to post the data (using `apollo-upload-client` and [ReactNativeFile](https://github.com/jaydenseric/apollo-upload-client#class-reactnativefile) based on file path) in asynchronous way to not force our users to get a network connection.

As the file is removed every time the user is picking another file, i tried to use the base64 content of the file, to store it and reuse it before upload, but as you mentioned it's clear enough that base64 is not a suitable solution.

So at this time, we are stuck because we only have 2 solutions:
- using 2.x version which is not maintained and doesn't support Android 10+ / iOS 14
- using 3.x version that removes temporary files each time we pick another photo

Is there any way we can disable automatic file removal from your lib with an option and/or a callback to be able to trigger the file removal manually once the upload has been done successfully ?

For example, i'd like to be able to define options like:
```js
{
  tempFileDelete: false,
}
```

and being able to trigger the deletion like:

```js
import { removeTempFile } from 'react-native-image-picker';

const handleFileUpload = () => {
  [...]
  removeTempFile(localPath);
};
```

Like [this](https://github.com/ivpusic/react-native-image-crop-picker#optional-cleanup) in few words

**Reply by ravirajn22** [reply]
Will something like holding to say latest 15 files and deleting all other will do? Instead of all these options.

Stick with 2.x.x for now, until a solution is reached

**Reply by VincentCATILLON** [reply]
Hey @ravirajn22 , 

Unfortunately, that won't fit our (and potentially some other people too) needs as we need to trigger the deletion manually when some events on our app has been triggered (like upload success).

**Reply by ravirajn22** [reply]
Latest version no longer uses auto-delete, it just uses cache folder and all image uri will persist sessions untill its cleared from cache folder by the OS.
